src/equations/ideal_glm_mhd_multiion_2d.jl

```diff
@@ -0,0 +1,859 @@
+# for all species and the maximum magnetosonic speed.
+@inline function max_abs_speed_naive(u_ll, u_rr, orientation::Integer,
+                                     equations::IdealGlmMhdMultiIonEquations2D)
+    # Calculate fast magnetoacoustic wave speeds
+    # left
+    cf_ll = calc_fast_wavespeed(u_ll, orientation, equations)
+    # right
+    cf_rr = calc_fast_wavespeed(u_rr, orientation, equations)
+
+    # Calculate velocities
+    v_ll = zero(eltype(u_ll))
+    v_rr = zero(eltype(u_rr))
+    if orientation == 1
+        for k in eachcomponent(equations)
+            rho, rho_v1, _ = get_component(k, u_ll, equations)
+            v_ll = max(v_ll, abs(rho_v1 / rho))
+            rho, rho_v1, _ = get_component(k, u_rr, equations)
+            v_rr = max(v_rr, abs(rho_v1 / rho))
+        end
+    else #if orientation == 2
+        for k in eachcomponent(equations)
+            rho, rho_v1, rho_v2, _ = get_component(k, u_ll, equations)
+            v_ll = max(v_ll, abs(rho_v2 / rho))
+            rho, rho_v1, rho_v2, _ = get_component(k, u_rr, equations)
+            v_rr = max(v_rr, abs(rho_v2 / rho))
+        end
+    end
+
+    λ_max = max(abs(v_ll), abs(v_rr)) + max(cf_ll, cf_rr)
+end
+
+@inline function max_abs_speeds(u, equations::IdealGlmMhdMultiIonEquations2D)
+    v1 = zero(real(equations))
+    v2 = zero(real(equations))
+    for k in eachcomponent(equations)
+        rho, rho_v1, rho_v2, _ = get_component(k, u, equations)
+        v1 = max(v1, abs(rho_v1 / rho))
+        v2 = max(v2, abs(rho_v2 / rho))
+    end
+
+    cf_x_direction = calc_fast_wavespeed(u, 1, equations)
+    cf_y_direction = calc_fast_wavespeed(u, 2, equations)
+
+    return (abs(v1) + cf_x_direction, abs(v2) + cf_y_direction)
+end
+
+# Compute the fastest wave speed for ideal multi-ion GLM-MHD equations: c_f, the fast 
+# magnetoacoustic eigenvalue. This routine computes the fast magnetosonic speed for each ion
+# species using the single-fluid MHD expressions and approximates the multi-ion c_f as 
+# the maximum of these individual magnetosonic speeds.
+@inline function calc_fast_wavespeed(cons, orientation::Integer,
+                                     equations::IdealGlmMhdMultiIonEquations2D)
+    B1, B2, B3 = magnetic_field(cons, equations)
+    psi = divergence_cleaning_field(cons, equations)
+
+    c_f = zero(real(equations))
+    for k in eachcomponent(equations)
+        rho, rho_v1, rho_v2, rho_v3, rho_e = get_component(k, cons, equations)
+
+        rho_inv = 1 / rho
+        v1 = rho_v1 * rho_inv
+        v2 = rho_v2 * rho_inv
+        v3 = rho_v3 * rho_inv
+        v_mag = sqrt(v1^2 + v2^2 + v3^2)
+        gamma = equations.gammas[k]
+        p = (gamma - 1) *
+            (rho_e - 0.5f0 * rho * v_mag^2 - 0.5f0 * (B1^2 + B2^2 + B3^2) -
+             0.5f0 * psi^2)
+        a_square = gamma * p * rho_inv
+        sqrt_rho = sqrt(rho)
+
+        b1 = B1 / sqrt_rho
+        b2 = B2 / sqrt_rho
+        b3 = B3 / sqrt_rho
```

Comment:
```suggestion
        inv_sqrt_rho = sqrt(rho)

        b1 = B1 * inv_sqrt_rho
        b2 = B2 * inv_sqrt_rho
        b3 = B3 * inv_sqrt_rho
```

Reply:
It should be `inv_sqrt_rho = 1 / sqrt(rho)`, not `inv_sqrt_rho = sqrt(rho)`.
Thanks for the tip though. Done in 373aab057db7bab53621bc1e570695d322823c02.